### PR TITLE
Parallelize arangorestore using the tools developed for arangodump

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

File: arangosh/Dump/DumpFeature.cpp [hunk 1]
```diff
@@ -209,6 +209,28 @@ void flushWal(arangodb::httpclient::SimpleHttpClient& client) {
 }
 }  // namespace
 
+namespace {
+bool isIgnoredHiddenEnterpriseCollection(
```
Comment (graetzer) [inline arangosh/Dump/DumpFeature.cpp]: Can we move this helper method somewhere more general ?

File: arangosh/Dump/DumpFeature.cpp [hunk 2]
```diff
@@ -209,6 +209,28 @@ void flushWal(arangodb::httpclient::SimpleHttpClient& client) {
...
+             "dump this collection anyway use 'arangodump --force'. "
+             "However this is not recommended and you should instead dump "
+             "the EdgeCollection of the SmartGraph instead.";
+      return true;
+    }
+  }
+#endif
+  return false;
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Dump/DumpFeature.cpp]: why?

File: arangosh/Dump/DumpFeature.cpp [hunk 3]
```diff
@@ -322,7 +343,8 @@ arangodb::Result handleCollection(
 
   // do the hard work in another function...
   return ::dumpCollection(client, jobData, file, jobData.name, "",
-                          jobData.batchId, jobData.tickStart, jobData.tickEnd);
+                          jobData.batchId, jobData.options.tickStart,
+                          jobData.options.tickEnd);
 }
 }  // namespace
```
Comment (ObiWahn) [inline arangosh/Dump/DumpFeature.cpp]: why?

Comment (ObiWahn) [reply arangosh/Dump/DumpFeature.cpp]: in 447 as well why do we close and immediately re-open the unnamed namespace?!

File: arangosh/Restore/RestoreFeature.cpp [hunk 4]
```diff
@@ -50,397 +44,494 @@
...
+            "got invalid response from server: " + client.getErrorMessage()};
+  }
+  if (response->wasHttpError()) {
+    return {TRI_ERROR_INTERNAL, "got invalid response from server: HTTP " +
+                                    itoa(response->getHttpReturnCode()) + ": " +
+                                    response->getHttpReturnMessage()};
+  }
+  return {TRI_ERROR_NO_ERROR};
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 5]
```diff
@@ -50,397 +44,494 @@
...
-  }
+  // Finally, sort by name so we have stable, reproducible results
+  std::string leftName =
+      arangodb::basics::VelocyPackHelper::getStringValue(left, "name", "");
+  std::string rightName =
+      arangodb::basics::VelocyPackHelper::getStringValue(right, "name", "");
+
+  return strcasecmp(leftName.c_str(), rightName.c_str()) < 0;
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 6]
```diff
@@ -50,397 +44,494 @@
...
   }
 
   // any other error
-  _httpClient->setErrorMessage(getHttpErrorMessage(response.get(), nullptr),
-                               false);
-  return TRI_ERROR_INTERNAL;
+  auto res = ::checkHttpResponse(*httpClient, response);
+  return {TRI_ERROR_INTERNAL, res.errorMessage()};
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 7]
```diff
@@ -50,397 +44,494 @@
...
+             "when invoking arangorestore with an encrypted dump";
+    } else {
+      LOG_TOPIC(INFO, Logger::RESTORE)
+          << "# using encryption type " << directory.encryptionType()
+          << " for reading dump";
+    }
+#endif
+  }
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 8]
```diff
@@ -50,397 +44,494 @@
...
+  if (forceSameDatabase && databaseName != client->databaseName()) {
+    return {TRI_ERROR_BAD_PARAMETER,
+            std::string("database name in dump.json ('") + databaseName +
+                "') does not match specified database name ('" +
+                client->databaseName() + "')"};
+  }
+
+  return {};
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 9]
```diff
@@ -50,397 +44,494 @@
...
-    }
-
-    return TRI_ERROR_INTERNAL;
+    return ::checkHttpResponse(httpClient, response);
   }
 
-  return TRI_ERROR_NO_ERROR;
+  return {TRI_ERROR_NO_ERROR};
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 10]
```diff
@@ -50,397 +44,494 @@
...
-                                    char const* buffer, size_t bufferSize,
-                                    std::string& errorMsg) {
-  std::string const url = "/_api/replication/restore-data?collection=" +
-                          StringUtils::urlEncode(cname) +
-                          "&force=" + (_force ? "true" : "false");
-
-  std::unique_ptr<SimpleHttpResult> response(
-      _httpClient->request(rest::RequestType::PUT, url, buffer, bufferSize));
-
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 11]
```diff
@@ -50,397 +44,494 @@
...
-    }
-
-    return TRI_ERROR_INTERNAL;
+    return ::checkHttpResponse(httpClient, response);
   }
 
-  return TRI_ERROR_NO_ERROR;
+  return {TRI_ERROR_NO_ERROR};
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 12]
```diff
@@ -50,397 +44,494 @@
...
 
-  if (leftDist.isNone() && !rightDist.isNone()) {
-    return true;
+  if (result.fail() && jobData.options.force) {
+    LOG_TOPIC(ERR, Logger::RESTORE) << result.errorMessage();
+    result.reset();
   }
+  return result;
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 13]
```diff
@@ -50,397 +44,494 @@
...
-      arangodb::basics::VelocyPackHelper::getNumericValue<int>(left, "type", 0);
-  int rightType = arangodb::basics::VelocyPackHelper::getNumericValue<int>(
-      right, "type", 0);
+  if (result.fail() && jobData.options.force) {
+    LOG_TOPIC(ERR, Logger::RESTORE) << result.errorMessage();
+    result.reset();
+  }
+  return result;
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: why?

File: arangosh/Restore/RestoreFeature.cpp [hunk 14]
```diff
@@ -50,397 +44,494 @@
...
-                      client->databaseName() + "')");
+    if (numRead == 0) {  // EOF
+      break;
+    }
   }
 
-  return Result();
+  return result;
 }
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: it is not like we end up in another namespace:P

File: arangosh/Restore/RestoreFeature.cpp [hunk 15]
```diff
@@ -524,185 +610,240 @@ int RestoreFeature::processInputDirectory(std::string& errorMsg) {
...
+    return {TRI_ERROR_INTERNAL,
+            std::string(
+                "arangorestore terminated because of an unhandled exception: ")
+                .append(ex.what())};
+  } catch (...) {
+    return {TRI_ERROR_OUT_OF_MEMORY, "arangorestore out of memory"};
+  }
+  return {TRI_ERROR_NO_ERROR};
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: you just want 40 lines extra

File: arangosh/Restore/RestoreFeature.cpp [hunk 16]
```diff
@@ -524,185 +610,240 @@ int RestoreFeature::processInputDirectory(std::string& errorMsg) {
...
+  if (!jobData.options.indexesFirst && jobData.options.importStructure) {
+    // restore indexes second if we are using mmfiles
+    result = ::restoreIndexes(httpClient, jobData);
+    if (result.fail()) {
+      return result;
+    }
+  }
+  return result;
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: 42

File: arangosh/Restore/RestoreFeature.cpp [hunk 17]
```diff
@@ -524,185 +610,240 @@ int RestoreFeature::processInputDirectory(std::string& errorMsg) {
...
+namespace {
+/// @brief handle the result of a single job
+void handleJobResult(
+    std::unique_ptr<arangodb::RestoreFeature::JobData>&& jobData,
+    arangodb::Result const& result) {
+  if (result.fail()) {
+    jobData->feature.reportError(result);
+  }
+}
+}  // namespace
```
Comment (ObiWahn) [inline arangosh/Restore/RestoreFeature.cpp]: :100:

**Comment by danielhlarkin** [comment]
@ObiWahn I've combined adjacent unnamed namespaces.

File: arangosh/Dump/DumpFeature.cpp (hunk 1 above)
Comment (danielhlarkin) [reply arangosh/Dump/DumpFeature.cpp]: Sure, do you have a suggestion? Maybe static method on `LogicalCollection`?

**Pull request merged**